### PR TITLE
Parser TS bindings

### LINKED ISSUE #7884: Exposing AST API through typescript

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Generate TS bindings and lazy deserialization for the parser types.

### Important Notes

- The new API is imported into `ffi.ts`, but not yet used.
- I have tested the generated code in isolation, but cannot commit tests as we are not currently able to load WASM modules when running in `vitest`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by farmaazon** [comment]
> I have tested the generated code in isolation, but cannot commit tests as we are not currently able to load WASM modules when running in vitest.

@kazcw I have very similar problem, but also have something like solution:
```ts
if (amIInTestMode()) {
  const fs = await import('node:fs/promises')
  const buffer = await fs.readFile('./rust-ffi/pkg/rust_ffi_bg.wasm')
  await init(buffer)
} else {
  await init()
}
```
(in place of simple `await init()` on the top of `ffi.ts`).
Where `amIInTestMode` should check if we're running in browser or in node. I haven't implemented it myself yet, as haven't decided [which way of detecting node is the best for our case](https://stackoverflow.com/questions/4224606/how-to-check-whether-a-script-is-running-under-node-js)

But you can just put if (true) and make tests on separate branch, once my current PR will fix the problem properly you can also make PR for parser tests.

File: app/gui2/src/util/ffi.ts [hunk 1]
```diff
@@ -9,6 +10,11 @@ export function parseEnso(code: string): Ast.Tree {
   return JSON.parse(json)
 }
 
+export function parseEnso2(code: string): Ast2.Tree {
+  const blob = parse(code)
+  return Ast2.deserializeTree(blob)
+}
+
```
Comment (farmaazon) [inline app/gui2/src/util/ffi.ts]: Please add TODO explaining that (and possibly when) we plan to remove old AST and rename parseEnso2 and Ast2 to normal parseEnso and Ast.

File: app/gui2/src/util/parserSupport.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  protected readonly lazyObjectData: Cursor
+
+  constructor(data: Cursor) {
+    this.lazyObjectData = data
+  }
+
+  debug(): {} {
+    return {}
+  }
+}
```
Comment (farmaazon) [inline app/gui2/src/util/parserSupport.ts]: I don't understand this class. Could you add a documentation?

File: app/gui2/src/util/parserSupport.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+export type Result<T, E> = { Ok: T } | { Err: E }
```
Comment (farmaazon) [inline app/gui2/src/util/parserSupport.ts]: You may check out the module I created on my branch: https://github.com/enso-org/enso/blob/wip/farmaazon/loading-suggestion-db/app/gui2/src/util/result.ts it has some additional useful stuff.

File: lib/rust/parser/src/format.rs [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+use std::fmt::Formatter;
+
+
+// =================
+// === Constants ===
+// =================
+
+/// Maximum allowed nesting depth of compound objects. This is empirically determined to be reached
+/// before stack overflow on supported targets.
+// FIXME: WASM test required.
```
Comment (farmaazon) [inline lib/rust/parser/src/format.rs]: When do we plan to fix this FIXME? If not immediately, then could you create an issue for it, and link it here?

File: lib/rust/parser/src/format.rs [hunk 5]
```diff
@@ -0,0 +1,515 @@
+//! # Design
+//!
+//! The format produced by this module is not actually "serial", as such there is a bit of an
+//! impedance mismatch using the `Serializer` trait: `serde` presents each field to the
+//! serializer once, but we ultimately need to write to two different places in the output for
+//! each "boxed" field (the data, and the reference to it).
+//!
+//! The approach used here is to maintain a stack of the fields of incomplete objects as we
+//! descend in to them; when an object is completed, it is moved to the heap. This requires
+//! moving each boxed object once.
```
Comment (farmaazon) [inline lib/rust/parser/src/format.rs]: Maybe it is obvious, but I would add a note why we need to serialize boxes this way: w want to serialize recursive structures.

BTW, I'm surprised there is no ready crate for such scenario. Maybe we should make it a separate crate and publish it?

File: lib/rust/parser/schema/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            field.offset = offset;
+            offset += reference_size
+                .get(&field.r#type)
+                .copied()
+                .unwrap_or_else(|| data_size[&field.r#type]);
+        }
+        let reference = TypeRef::Type { id: id.to_owned() };
+        ty.size = reference_size.get(&reference).copied().unwrap_or_else(|| data_size[&reference]);
+    }
+    Schema { types }
```
Comment (farmaazon) [inline lib/rust/parser/schema/src/lib.rs]: This last part is obscure a bit, it is hard to guess what role it plays.

You can add comments as in the sections before, but there's a better approach: splitting it into a set of well-named functions. You may create a "Builder" structure with all mutable variables here as fields, and make each block a method of it. And `compute_size` could be a method of such structure as well.

File: app/gui2/parser-codegen/src/codegen.ts [hunk 7]
```diff
@@ -0,0 +1,694 @@
+import * as fs from 'fs'
+import * as ts from 'typescript'
+import { factory as tsf } from 'typescript'
+
+type DiscriminantMap = {
+  [discriminant: number]: string
+}
```
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts]: What are the values of this map?

File: app/gui2/parser-codegen/src/codegen.ts [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+type TypeGraph = {
+  [id: string]: Schema.Type
+}
```
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts]: Isn't `TypeGraph` a part of `Schema`?

File: app/gui2/parser-codegen/src/codegen.ts [hunk 9]
```diff
@@ -0,0 +1,694 @@
...
+          readElement(innerParameter),
+        ),
+      ],
+    )
+  }
+}
+
+function readerTransformer2(
+  name: string,
+): (readOk: ExpressionTransformer, readErr: ExpressionTransformer) => ExpressionTransformer {
```
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts]: What `2` means here? Please rename.

File: app/gui2/parser-codegen/src/codegen.ts [hunk 10]
```diff
@@ -0,0 +1,694 @@
...
+type ExpressionTransformer = (expression: ts.Expression) => ts.Expression
+
+function primitiveReader(name: string): ExpressionTransformer {
+  return (cursor) =>
+    tsf.createCallExpression(tsf.createPropertyAccessExpression(cursor, name), [], [])
+}
+
+function readerTransformer(
+  name: string,
+): (readElement: ExpressionTransformer) => ExpressionTransformer {
```
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts]: Could you add a short description here, what exactly this function returns? What logically the returned transformer accepts and what produces?

File: lib/rust/parser/src/format.rs (hunk 4 above)
Comment (kazcw) [reply lib/rust/parser/src/format.rs]: Fixed

File: app/gui2/src/util/ffi.ts [hunk 11]
```diff
@@ -523,6 +530,25 @@ export namespace Doc {
...
+    if (tree0.type === Ast2.Tree.Type.BodyBlock) {
+      const statements = Array.from(tree0.statements)
+      const tree1 = statements[0]!.expression!
+      expect(tree1.type).toBe(Ast2.Tree.Type.Ident)
+      if (tree1.type === Ast2.Tree.Type.Ident) {
+        expect(Number(tree1.token.codeUtf16)).toBe(identInput.length)
+      }
+    }
+  })
+}
```
Comment (farmaazon) [inline app/gui2/src/util/ffi.ts]: Please add more tests for cases like operator application, block (operator) application, application with default argument, erroneous AST and possibly more. This is a core part of almost every operation we do, and I'd like to be fairy sure it works for basic scenarios. The tests may be in a separate file (we already have some in `__tests__` directory).

File: lib/rust/parser/schema/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+//! Generate a schema representing `enso-parser`'s AST types. This schema can be used to generate
+//! AST representations and (de)serialization.
```
Comment (farmaazon) [inline lib/rust/parser/schema/src/main.rs]:
```suggestion
//! Generate a schema representing `enso-parser`'s AST types. This schema can be used to generate
//! AST representations in other languages and (de)serialization.
```

File: lib/rust/parser/src/format.rs (hunk 5 above)
Comment (farmaazon) [reply lib/rust/parser/src/format.rs]: > Maybe it is obvious, but I would add a note why we need to serialize boxes this way: w want to serialize recursive structures.

Well, the description is needed even more, given that what I wrote before is actually wrong :) We do such boxing to have structures equal in size, what helps to implement "lazy deserialization" in js.

File: app/gui2/parser-codegen/src/codegen.ts [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+    .join('')
+}
+
+function toPascal(ident: string, prefix?: string): string {
+  return fromSnake(ident, 'pascal', prefix)
+}
+
+function toCamel(ident: string, prefix?: string): string {
+  return fromSnake(ident, 'camel', prefix)
+}
```
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts]: I would add simple tests for those functions.

File: app/gui2/parser-codegen/src/codegen.ts
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts:1]: This file is quite big. Can we split it into submodules? For example:
* One with helper functions (like toPascal) with simple inline tests for those
* One for code generating functions (also with tests if possible)
* Optionally one for schema types 
* And one for "Main" code.

Also, I miss a description of what code this actually produces, and why we're doing it this way - after all, this is not a simple TS structures representing AST code.

File: app/gui2/parser-codegen/src/codegen.ts [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+  return (cursor) =>
+    tsf.createCallExpression(tsf.createPropertyAccessExpression(cursor, name), [], [])
+}
+
+/**
+ * Given the name of a runtime `Cursor` method that deserializes a derived type given a function to deserialize a
+ * base type, return a codegen-time function that generates a *reader* for a derived type from a *reader* for the base
+ * type, where a *reader* is a function producing a deserialization expression from an expression that evaluates to a
+ * `Cursor`.
+ */
```
Comment (farmaazon) [inline app/gui2/parser-codegen/src/codegen.ts]: An example would make it greatly easier to understand this function.

File: lib/rust/parser/src/format.rs [hunk 15]
```diff
@@ -0,0 +1,529 @@
...
+            fn new() -> Rc<Self> {
+                let cyclic = Rc::new(Cyclic { this: RefCell::new(None) });
+                *cyclic.this.borrow_mut() = Some(Rc::clone(&cyclic));
+                cyclic
+            }
+        }
+        // Note that if recursion is not adequately limited the expected failure mode is aborting
+        // due to stack overflow. We are just checking `is_err` here for good measure.
+        assert!(super::serialize(Cyclic::new()).is_err());
+    }
```
Comment (farmaazon) [inline lib/rust/parser/src/format.rs:537]: Add a test case for serializing some simple structure, and also simple but with boxed fields structure (checking the generated binary content).

**Review by farmaazon — CHANGES_REQUESTED** [review]
The parser js API should have much more automatic tests. The documentation also should be improved, but it's not a blocker for merging if it would take too much time.

**Comment by farmaazon** [comment]
For example: This test created on my branch derived from this one is raising an exception:
```ts
  test('TryParseSmthing', () => {
    const smthing = 'Data.read file="name"'
    const parsed = parseEnso2(smthing)
    Array.from(parsed.statements)[0].expression.name.isTypeOrConstructor
  })
```

File: app/gui2/parser-codegen/src/codegen.ts (hunk 13 above)
Comment (Frizi) [reply app/gui2/parser-codegen/src/codegen.ts]: We should really use a dependency for this kind of stuff. Having to write and test it ourselves is really pointless. There is a good package for that: https://www.npmjs.com/package/change-case

**Review by farmaazon — APPROVED** [review]

**Review by Frizi — APPROVED** [review]

**Pull request merged**